server/src/main/java/org/elasticsearch/cluster/coordination/ClusterFormationFailureHelper.java

```diff
@@ -249,5 +254,37 @@ private static String describeQuorum(VotingConfiguration votingConfiguration) {
                 }
             }
         }
+
+        private String getJoinStatusDescription() {
+            if (inFlightJoinStatuses.isEmpty()) {
+                return "";
+            }
+
+            final var stringBuilder = new StringBuilder();
+            inFlightJoinStatuses.stream()
```

Comment:
Nit: I think it would be possible to collect results to the `stringBuilder` directly with the Streams API
```java
return inFlightJoinStatuses.stream()
                .sorted(Comparator.comparing(JoinStatus::age).reversed())
                .limit(10)
                .collect(
                    StringBuilder::new,
                    (sb, joinStatus) -> sb.append("; joining [")
                        .append(joinStatus.remoteNode().descriptionWithoutAttributes())
                        .append("] in term [")
                        .append(joinStatus.term())
                        .append("] has status [")
                        .append(joinStatus.message())
                        .append("] after [")
                        .append(timeValueWithMillis(joinStatus.age()))
                        .append("]"),
                    StringBuilder::append
                )
                .toString();
```

Reply:
I think that's right, but I don't like this pattern because the execution order is poorly defined and doesn't exclude the possibility of doing extra unnecessary copying. It's cheap to append a few characters to the end of a `StringBuilder` but we never want to append a whole `StringBuilder`.

Not that it really matters here, there's only 10 items and this isn't exactly hot code, but still I don't like the pattern.